{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97ce68b4-1fe2-450b-b80d-a140e68d3191",
   "metadata": {},
   "source": [
    "# Calling genotypes, prevalences and filtering data\n",
    "This program analyzes tables of counts. In the prior \"variant calling\" step of the program, mutation counts were generated for each sample by annotating the VCF file generated by freebayes, using SNPEff to assign amino acid names to mutations.\n",
    "\n",
    "The count tables are intended to be as \"raw\" as possible. In this part of the analysis we will:\n",
    "1. apply various filters to the count tables, considering a mutation 'sampled' for a sample if it passes the filters for that sample.\n",
    "2. choose the level of detail of geographic locations we care about from a metadata file\n",
    "3. summarize the number of filtered samples at each location that had the mutation and the total number of filtered samples at each location, where the 'prevalence' of a mutation at a location is the first number divided by the second number\n",
    "4. graph the prevalences and filtered sample sizes on a map\n",
    "5. (optional) - apply additional filtering for re-evaluation of prevalence data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efb7fc2e-61e7-44e4-aeff-8fa181fc25ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit\n",
    "output_folder = 'output'\n",
    "mutation_count_file = \"/home/charlie/git/seekdeep_amplicon_visualization/input/new_tables/alternate_AA_table.csv\"\n",
    "mutation_coverage_file = \"/home/charlie/git/seekdeep_amplicon_visualization/input/new_tables/coverage_AA_table.csv\"\n",
    "metadata_table = \"/home/charlie/git/seekdeep_amplicon_visualization/input/new_tables/randomly_sampled_metadata.tsv\"\n",
    "\n",
    "# how many reads (or, more accurately, UMIs) do we need spanning the genomic coordinates of a mutation in order to\n",
    "# consider the coordinates \"sequenced\" for a given sample? If UMI count is below min_coverage, the sample will be\n",
    "# marked as \"missing\" for the mutation of interest.\n",
    "min_coverage = 3\n",
    "\n",
    "# If min_coverage is achieved in the sample, how many UMIs do we need that support the mutant (or alternate) allele\n",
    "# in order to consider the alternate allele \"present\" in the sample? Otherwise if min_coverage is surpassed but\n",
    "# min_count is not, the sample will be marked as matching the reference allele.\n",
    "min_count = 1\n",
    "\n",
    "# Is there a number that min_count/min_coverage needs to surpass before we consider an alternate allele \"present\" in a\n",
    "# sample? Our usual answer is 'no' so we generally set this to '0' - but you may not want to count an alternate allele\n",
    "# present if it's found in a very low fraction of the UMIs associated with a sample, even if the absolute number of UMIs\n",
    "# associated with the alternate allele is above your threshold.\n",
    "min_freq = 0\n",
    "\n",
    "# Don't Edit below\n",
    "import sys\n",
    "sys.path.append(\"src\")\n",
    "import background as bg\n",
    "all_columns, mutation_dict, filtered_mutations, mutation_counts, mutation_coverage = bg.get_mutation_counts(mutation_count_file, mutation_coverage_file)\n",
    "print('observed missense mutations are:')\n",
    "print(filtered_mutations)\n",
    "print('')\n",
    "metadata_columns = bg.get_metadata_columns(metadata_table)\n",
    "print('metadata columns are: ')\n",
    "print(metadata_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7283eb2-cdb9-449f-9577-1ee4c48274a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which mutations would you like to graph? A full list of sequenced plus key mutations from targets.tsv is created when\n",
    "#you run the code in the cell above - this outputs the full list of available mutations to choose from.\n",
    "\n",
    "#example if you want only \"key mutations from the targets.tsv file found in the project_resources folder\" plus k13 mutations\n",
    "#mutation_column=open('/opt/project_resources/targets.tsv').readline().strip().split('\\t').index('mutation_name')\n",
    "#mutations_of_interest=[line.strip().split('\\t')[mutation_column] for line in open('/opt/project_resources/targets.tsv')][1:]\n",
    "#k13_muts=[mut for mut in filtered_mutations if 'k13' in mut]\n",
    "#mutations_of_interest+=['dhps-Ala437Gly'] #targets.tsv file seems to be missing this key mutation\n",
    "#mutations_of_interest+=k13_muts #add any additional found k13 mutations\n",
    "#mutations_of_interest=list(set(mutations_of_interest)) #remove any duplicates\n",
    "\n",
    "#example if you want only mutations from the AA tables from a specific gene\n",
    "#mutations_of_interest=[mutation for mutation in filtered_mutations if 'k13' in mutation]\n",
    "\n",
    "#example if you want to manually set the mutations to analyze\n",
    "# mutations_of_interest=[\n",
    "# 'dhfr-ts-Asn51Ile', 'dhfr-ts-Cys59Arg', 'dhfr-ts-Ser108Asn', 'dhfr-ts-Ile164Leu',\n",
    "# 'mdr1-Asn652Asp', 'mdr1-Asn86Tyr', 'mdr1-Asp1246Tyr', 'mdr1-Asp650Asn',\n",
    "# 'mdr1-Tyr184Phe', 'crt-Met74Ile', 'crt-Asn75Glu', 'crt-Lys76Thr',\n",
    "# 'crt-Gln271Glu', 'crt-Arg371Ile', 'dhps-Ala437Gly', 'dhps-Lys540Glu',\n",
    "# 'dhps-Ala581Gly', 'k13-Arg561His']\n",
    "\n",
    "mutations_of_interest=[\n",
    "    'crt-Lys76Thr',\n",
    "    'k13-Arg561His',\n",
    "    'dhfr-ts-Ile164Leu',\n",
    "    'dhps-Ala581Gly',\n",
    "    'mdr1-Asn86Tyr'\n",
    "]\n",
    "\n",
    "# In the metadata sheet (reproduced in the table output from the cell above), which column contains sample names and\n",
    "# which column contains the names of the locations you'd like to plot (or summarize by)?\n",
    "sample_column='Sample_ID'\n",
    "summary_column='Region'\n",
    "\n",
    "# don't edit\n",
    "bg.create_prevalences_input_table(mutations_of_interest, mutation_dict, all_columns, mutation_counts, mutation_coverage, output_folder, min_count, min_coverage, min_freq)\n",
    "prevalences = bg.calculate_prevalences(output_folder,metadata_table, sample_column, summary_column, mutations_of_interest)\n",
    "prevalences.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1aef84d-8f44-4a68-8e2d-b01f53b450c0",
   "metadata": {},
   "source": [
    "# Plot Scatter map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee5b472f-25b7-4e97-b082-7b761eba7677",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads a dropdown menu so a user can choose one mutation of interest to graph\n",
    "# RUN\n",
    "import os\n",
    "import ipywidgets as widgets\n",
    "variants_of_interest=open(os.path.join(output_folder, \"prevalence_summary.tsv\")).readline().strip().split('\\t')[3:]\n",
    "variant = widgets.Dropdown(\n",
    "    options=variants_of_interest,\n",
    "    description='variant:',\n",
    "    disabled=False,\n",
    ")\n",
    "# Don't edit\n",
    "country, country_shortcuts = bg.generate_country_dropdown()\n",
    "print(\"Country and Variant of Interest:\\n\")\n",
    "display(country)\n",
    "display(variant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e2651e2-e329-4bb6-9a21-7ab1077fc86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reads in a country using the dropdown menu (above) or the (currently commented out) manual values below\n",
    "# RUN\n",
    "country=country.value\n",
    "zoom_level=country_shortcuts[country]['zoom']\n",
    "latitude=country_shortcuts[country]['lat']\n",
    "longitude=country_shortcuts[country]['lon']\n",
    "# OPTIONAL USER INPUT:\n",
    "# You can override the values above and set a custom zoom level and centering below by commenting these lines in\n",
    "#zoom_level=4.8\n",
    "#latitude=1.5\n",
    "#longitude=32\n",
    "\n",
    "# graphs the mutation of interest\n",
    "fig = bg.make_detail_graph(variant.value, summary_column, output_folder, zoom_level, latitude, longitude)\n",
    "fig.show()\n",
    "title_string=f'{variant.value}_{summary_column}'\n",
    "fig.write_image(os.path.join(output_folder, title_string+'.svg'))\n",
    "fig.write_image(os.path.join(output_folder, title_string+'.png'))\n",
    "fig.write_html(os.path.join(output_folder, title_string+'.html'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67929073-b47b-44e2-b981-c84a38afed06",
   "metadata": {},
   "source": [
    "## Static map of prevalences\n",
    "Creates an SVG file of mapped mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b81a9e9-1007-4b81-9e4f-bfd935085155",
   "metadata": {},
   "outputs": [],
   "source": [
    "svg_country = bg.get_countries_from_geojson()\n",
    "display(svg_country)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceebf9e5-8fed-4b02-99d1-3c8d98cefe9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit\n",
    "\n",
    "graph_states_provinces = True\n",
    "latitude_range = [-1.6, 4.4]\n",
    "longitude_range = [29, 35]\n",
    "scale_factor=1\n",
    "\n",
    "# additional labels to put on graph\n",
    "labels = [\n",
    "    ['Rwanda', -1.84, 30],\n",
    "    ['Kenya', 0.93, 35.1],\n",
    "    ['Tanzania', -1.75, 34.2],\n",
    "    ['Ethiopia', 4.8, 37.4],\n",
    "    ['S. Sudan', 4.16, 32.4],\n",
    "    ['DRC', 1.5, 29]\n",
    "]\n",
    "annotation_font_size = 8\n",
    "\n",
    "# title text\n",
    "title_text = variant.value+' Prevalence'\n",
    "title_size = 20\n",
    "\n",
    "# Don't Edit\n",
    "fig = bg.create_static_maps(\n",
    "    variant.value,\n",
    "    svg_country.value,\n",
    "    graph_states_provinces,\n",
    "    latitude_range,\n",
    "    longitude_range,\n",
    "    scale_factor,\n",
    "    labels,\n",
    "    annotation_font_size,\n",
    "    title_text,\n",
    "    title_size,\n",
    "    output_folder,\n",
    "    summary_column\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf610f5d-9b26-453c-9dd6-4f3052080eea",
   "metadata": {},
   "source": [
    "## Bar Chart\n",
    "Creates a Bar chart of mutation prevalences in the overall dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf3aed95-6c6e-45ac-85dd-df970d46950f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "# This code generates a bar chart for your mutations of interest\n",
    "def get_prevalences(prevalence_file, mutation_list):\n",
    "\tprevalence_dict={'mutations':[], 'prevalences':[], 'gene_name':[]}\n",
    "\th_dict={}\n",
    "\tfor line_number, line in enumerate(open(prevalence_file)):\n",
    "\t\tline=line.strip().split('\\t')\n",
    "\t\tif line_number==0:\n",
    "\t\t\tfor column_number, column in enumerate(line):\n",
    "\t\t\t\th_dict[column]=column_number\n",
    "\t\telif line[0]=='overall':\n",
    "\t\t\tfor mutation in mutation_list:\n",
    "\t\t\t\tif mutation in h_dict:\n",
    "\t\t\t\t\tcolumn=h_dict[mutation]\n",
    "\t\t\t\t\tprevalence=float(line[column].split(' ')[0])\n",
    "\t\t\t\t\tprevalence_dict['mutations'].append(mutation)\n",
    "\t\t\t\t\tprevalence_dict['prevalences'].append(prevalence)\n",
    "\t\t\t\t\tprevalence_dict['gene_name'].append(mutation.split('-')[0])\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tprint(mutation, f'is not found in {prevalence_file}!!')\n",
    "\treturn prevalence_dict\n",
    "\n",
    "def make_bar_chart(prevalence_dict, graph_type=\"bar chart of variants_of_interest\", y_max=1.0):\n",
    "\timport pandas as pd\n",
    "\timport plotly.express as px\n",
    "\tdf = pd.DataFrame(prevalence_dict)\n",
    "\tfig1=px.bar(df, x='mutations', y='prevalences', color='gene_name', text_auto=False)\n",
    "\tfig1['layout']['xaxis']['title']=graph_type\n",
    "\tfig1['layout']['yaxis']['title']='overall prevalence'\n",
    "\tfig1.update_layout(yaxis=dict(range=[0, y_max]))\n",
    "\tfig1.write_html(f'{graph_type}.html')\n",
    "\tfig1.write_image(f'{graph_type}.svg')\n",
    "\tfig1.write_image(f'{graph_type}.png')\n",
    "\treturn fig1\n",
    "chart_name='output/drug resistance mutations'\n",
    "y_max=1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa59382c-dba4-40fa-94ad-21611cee8d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTIONAL USER INPUT\n",
    "# User can optionally reset the variants_of_interest to graph, chart_name, and y_max values (e.g. by commenting in one\n",
    "#of the blocks of lines below), otherwise the values will be pulled from the cells above\n",
    "\n",
    "#variants_of_interest=['crt-Met74Ile', 'crt-Asn75Glu', 'crt-Lys76Thr','crt-Gln271Glu',\n",
    "#'crt-Arg371Ile', 'dhfr-ts-Asn51Ile', 'dhfr-ts-Cys59Arg', 'dhfr-ts-Ser108Asn',\n",
    "#'dhfr-ts-Ile164Leu', 'dhps-Ala437Gly', 'dhps-Lys540Glu', 'dhps-Ala581Gly',\n",
    "#'mdr1-Asn86Tyr', 'mdr1-Tyr184Phe', 'mdr1-Asp650Asn', 'mdr1-Asn652Asp',\n",
    "#'mdr1-Asp1246Tyr']\n",
    "#chart_name='non-k13 mutations'\n",
    "#y_max=1.0\n",
    "\n",
    "#variants_of_interest=['k13-Lys108Glu', 'k13-His136Asn', 'k13-Lys189Asn', 'k13-Lys189Thr',\n",
    "#'k13-Arg255Lys', 'k13-Leu258Met', 'k13-Gln271His', 'k13-Cys469Phe',\n",
    "#'k13-Cys469Tyr', 'k13-Asn490Thr', 'k13-Arg561His', 'k13-Ala578Ser',\n",
    "#'k13-Ala675Val']\n",
    "#chart_name='k13 mutations'\n",
    "#y_max=0.5\n",
    "\n",
    "\n",
    "# NO USER INPUT BELOW\n",
    "prevalence_dict=get_prevalences('output/prevalence_summary.tsv', variants_of_interest)\n",
    "fig1=make_bar_chart(prevalence_dict, chart_name, y_max)\n",
    "fig1.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
